[lldb/Target] Add BorrowedStackFrame and make StackFrame methods virtual

medismailben opened the pull request:
This change makes StackFrame methods virtual to enable subclass overrides and introduces BorrowedStackFrame, a wrapper that presents an existing StackFrame with a different frame index.

This enables creating synthetic frame views or renumbering frames without copying the underlying frame data, which is useful for frame manipulation scenarios.

This also adds a new borrowed-info format entity to show what was the original frame index of the borrowed frame.

jimingham commented on lldb/include/lldb/Target/BorrowedStackFrame.h:
```diff
@@ -0,0 +1,193 @@
+
+  bool IsInlined() override { return m_borrowed_frame_sp->IsInlined(); }
+
+  bool IsSynthetic() const override {
```
Would we ever want an IsBorrowed to go along with all these other qualifiers?

JDevlieghere commented on lldb/include/lldb/Target/BorrowedStackFrame.h:
```diff
@@ -0,0 +1,193 @@
+  /// \param [in] override_frame_index
+  ///   The frame index to report instead of the borrowed frame's index.
+  BorrowedStackFrame(const lldb::StackFrameSP &borrowed_frame_sp,
+                     uint32_t override_frame_index);
```
Maybe `new_frame_index`? `override_frame_index` sounds like a boolean, and `overridden` is ambiguous as to whether it's the new or old index.

medismailben commented on lldb/source/Target/BorrowedStackFrame.cpp:
```diff
@@ -0,0 +1,176 @@
+  m_new_frame_index = index;
+}
+
+uint32_t BorrowedStackFrame::GetConcreteFrameIndex() const {
```
Right, I guess that's the best we can do with what we have.

jimingham commented on lldb/source/Target/BorrowedStackFrame.cpp:
```diff
@@ -0,0 +1,183 @@
+}
+
+uint32_t BorrowedStackFrame::GetConcreteFrameIndex() {
+  // Inline frames don't have their own concrete frame index.
```
That comment is not right.  Inline frames have the concrete frame index of the unwind frame in which they were inlined.  A better comment would be:

```
// FIXME: We need to find where the concrete frame into which this frame was inlined 
  // landed in the new stack frame list as that is the correct concrete frame index in this
  //stack frame.

```
But I also don't think you want to actually check IsInlined here.  Rather, when you set the m_new_concrete_frame_index in the constructor, you want to do:

```
new_concrete_frame_index
                                     ? *new_concrete_frame_index
                                     : IsInlined() ? LLDB_INVALID_ID : new_frame_index

```
then GetConcreteFrameIndex can just report whatever got set.

JDevlieghere commented on lldb/source/Core/FormatEntity.cpp:
```diff
@@ -1761,6 +1764,23 @@ bool FormatEntity::Format(const Entry &entry, Stream &s,
+      if (StackFrame *frame = exe_ctx->GetFramePtr()) {
+        if (BorrowedStackFrame *borrowed_frame =
+                llvm::dyn_cast<BorrowedStackFrame>(frame)) {
+          ;
```
```suggestion
```

JDevlieghere reviewed (commented):
LGTM if Jim's happy.

medismailben committed:
[lldb/Target] Add BorrowedStackFrame and make StackFrame methods virtual

This change makes StackFrame methods virtual to enable subclass overrides
and introduces BorrowedStackFrame, a wrapper that presents an existing
StackFrame with a different frame index.

This enables creating synthetic frame views or renumbering frames without
copying the underlying frame data, which is useful for frame manipulation
scenarios.

This also adds a new borrowed-info format entity to show what was the
original frame index of the borrowed frame.

Signed-off-by: Med Ismail Bennani <ismail@bennani.ma>

jimingham reviewed (approved):
LGTM
